video/clip.go

```diff
@@ -89,67 +97,109 @@ func ConvertUnixMillisToSeconds(requestID string, firstSegment *m3u8.MediaSegmen
+			return nil, []ClipSegmentInfo{}, fmt.Errorf("error clipping: failed to get a starting index: %w", err)
 		}
 	}
 
 	// Find the segment index that correlates with the specified endTime.
 	if endTime > manifestDuration {
-		endSegIdx = manifestSegments - 1
+		lastSegmentIdx := manifestSegments - 1
+		lastSegmentDuration := manifest.Segments[lastSegmentIdx].Duration
+		clipEndSegmentInfo = ClipSegmentInfo{SequenceID: lastSegmentIdx, Duration: lastSegmentDuration, ClipOffsetSecs: lastSegmentDuration}
 	} else {
-		endSegIdx, err = getRelevantSegment(manifest.Segments, endTime, startSegIdx)
+		clipEndSegmentInfo, err = getRelevantSegment(manifest.Segments, endTime, 0)
 		if err != nil {
-			return nil, fmt.Errorf("error clipping: failed to get an ending index")
+			return nil, []ClipSegmentInfo{}, fmt.Errorf("error clipping: failed to get an ending index")
 		}
 	}
 
+	firstSegmentToClip := clipStartSegmentInfo.SequenceID
+	lastSegmentToClip := clipEndSegmentInfo.SequenceID
 	// Generate a slice of all segments that overlap with startTime to endTime
-	relevantSegments := manifest.Segments[startSegIdx : endSegIdx+1]
+	relevantSegments := manifest.Segments[firstSegmentToClip : lastSegmentToClip+1]
 	totalRelSegs := len(relevantSegments)
 	if totalRelSegs == 0 {
-		return nil, fmt.Errorf("error clipping: no relevant segments found in the specified time range")
+		return nil, []ClipSegmentInfo{}, fmt.Errorf("error clipping: no relevant segments found in the specified time range")
 	}
 
-	log.Log(requestID, "Clipping segments", "from", startSegIdx, "to", endSegIdx)
+	log.Log(requestID, "Clipping segments", "from", firstSegmentToClip, "to", lastSegmentToClip)
 
-	return relevantSegments, nil
+	// If the clip start/end times fall within the same segment, then
+	// save only the single segment's info
+	var cs []ClipSegmentInfo
+	if firstSegmentToClip == lastSegmentToClip {
+		cs = []ClipSegmentInfo{clipStartSegmentInfo}
+	} else {
+		cs = []ClipSegmentInfo{clipStartSegmentInfo, clipEndSegmentInfo}
+	}
+	return relevantSegments, cs, nil
 }
 
-// Clips a segment by re-encoding so that I-frames are placed at the right intervals
-// Currently using these settings for the encode step:
+// Clips a segment from the specified start/end times. In some cases, the segment
+// will be re-encoded so that I-frames are placed at the right intervals.
+// This allows for frame-accurate clipping.
+// Currently using a combination of these settings for the encode step:
 //
 //	"c:v": "libx264": Specifies H.264 video codec.
-//	"r": "24": Sets video frame rate to 24 FPS.
 //	"g": "48": Inserts keyframe every 48 frames (GOP size).
 //	"keyint_min": "48": Minimum keyframe interval.
 //	"sc_threshold": 50: Detects scene changes with threshold 50.
 //	"bf": "0": Disables B-frames for bidirectional prediction.
-//	"c:a": "copy": Copies audio stream without re-encoding.
+//	"c:a": "aac": re-encode audio and clip.
```

Comment:
Why did this change?

Reply:
This was a mistake in copy/paste from local cmd line prototype to code. The audio track needs to be re-encoded so that it gets clipped at the same places as video track.